services/worker/src/worker/job_runners/split/opt_in_out_urls_scan_from_streaming.py

```diff
@@ -0,0 +1,356 @@
+    JobRunnerError,
+    get_previous_step_or_raise,
+)
+from worker.job_runners._datasets_based_job_runner import DatasetsBasedJobRunner
+from worker.utils import SplitFirstRowsResponse, get_rows_or_raise
+
+SplitOptInOutUrlsScanJobRunnerErrorCode = Literal[
+    "InfoError",
+    "TooManyColumnsError",
+    "PreviousStepStatusError",
+    "PreviousStepFormatError",
+    "MissingSpawningTokenError",
+    "ExternalServerError",
+]
+
+
+class SplitOptInOutUrlsScanJobRunnerError(JobRunnerError):
+    """Base class for exceptions in this module."""
+
+    def __init__(
+        self,
+        message: str,
+        status_code: HTTPStatus,
+        code: SplitOptInOutUrlsScanJobRunnerErrorCode,
+        cause: Optional[BaseException] = None,
+        disclose_cause: bool = False,
+    ):
+        super().__init__(
+            message=message, status_code=status_code, code=code, cause=cause, disclose_cause=disclose_cause
+        )
+
+
+class InfoError(SplitOptInOutUrlsScanJobRunnerError):
+    """Raised when the info could not be fetched."""
+
+    def __init__(self, message: str, cause: Optional[BaseException] = None):
+        super().__init__(message, HTTPStatus.INTERNAL_SERVER_ERROR, "InfoError", cause, True)
+
+
+class TooManyColumnsError(SplitOptInOutUrlsScanJobRunnerError):
+    """Raised when the dataset exceeded the max number of columns."""
+
+    def __init__(self, message: str, cause: Optional[BaseException] = None):
+        super().__init__(message, HTTPStatus.INTERNAL_SERVER_ERROR, "TooManyColumnsError", cause, True)
+
+
+class PreviousStepStatusError(SplitOptInOutUrlsScanJobRunnerError):
+    """Raised when the previous step gave an error. The job should not have been created."""
+
+    def __init__(self, message: str, cause: Optional[BaseException] = None):
+        super().__init__(message, HTTPStatus.INTERNAL_SERVER_ERROR, "PreviousStepStatusError", cause, False)
+
+
+class PreviousStepFormatError(SplitOptInOutUrlsScanJobRunnerError):
+    """Raised when the content of the previous step has not the expected format."""
+
+    def __init__(self, message: str, cause: Optional[BaseException] = None):
+        super().__init__(message, HTTPStatus.INTERNAL_SERVER_ERROR, "PreviousStepFormatError", cause, False)
+
+
+class MissingSpawningTokenError(SplitOptInOutUrlsScanJobRunnerError):
+    """Raised when the spawning.ai token is not set."""
+
+    def __init__(self, message: str, cause: Optional[BaseException] = None):
+        super().__init__(message, HTTPStatus.INTERNAL_SERVER_ERROR, "MissingSpawningTokenError", cause, False)
+
+
+class ExternalServerError(SplitOptInOutUrlsScanJobRunnerError):
+    """Raised when the spawning.ai server is not responding."""
+
+    def __init__(self, message: str, cause: Optional[BaseException] = None):
+        super().__init__(message, HTTPStatus.INTERNAL_SERVER_ERROR, "ExternalServerError", cause, False)
+
+
+class OptUrl(TypedDict):
+    url: str
+    row_idx: int
+    column_name: str
```

Comment:
it's useful to have the row_idx and column_name, at one point we might be able to show that a URL is opted-in or opted-out in the dataset viewer.

But it would be interesting to be able to ask for only a range of rows (as in /rows). This means that, if we want to do that, we would want to store the data in a format that can later be queried (another parquet file?)

Reply:
With @AndreaFrancis we were thinking of having one cache entry for every dataset chunk of 100k rows. But indeed scanning the full dataset and storing the opt-out URLs parquet files somewhere is also a good idea

Reply:
Yeah, maybe having it stored somewhere else will avoid having size concerns even with small chunks (mongodb limitation)

Reply:
btw (out of the scope of this PR): yesterday I was thinking that the `content` field could live outside of the database (on a filesystem, in S3, ...) Because: 1. it takes most of the space (and thus most of the costs of the database?), 2. it's not useful to have it in the database since we don't do queries over its content.